File: backend/app/schemas/analysis_comparison.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# Analysis Comparison Schemas
class AnalysisComparisonBase(BaseModel):
    name: str = Field(..., description="User-defined comparison name")
    description: str | None = Field(None, description="Optional comparison description")
    analysis_ids: list[int] = Field(..., description="List of analysis IDs to compare")
    comparison_type: str = Field(..., description="Type of comparison: side_by_side, temporal_trend, provider_performance")
    comparison_criteria: dict[str, Any] | None = Field(None, description="Comparison parameters and filters")


class AnalysisComparisonCreate(AnalysisComparisonBase):
    pass


class AnalysisComparisonUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    comparison_criteria: dict[str, Any] | None = None
    is_shared: bool | None = None


class AnalysisComparison(AnalysisComparisonBase):
    id: str
    user_id: int
    comparison_results: dict[str, Any] | None = None
    key_differences: dict[str, Any] | None = None
    statistical_insights: dict[str, Any] | None = None
    trend_analysis: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime
    is_shared: bool

    class Config:
        from_attributes = True


# Provider Performance Metrics Schemas
class ProviderPerformanceMetricsBase(BaseModel):
    provider_id: str
    period_start: datetime
    period_end: datetime
    period_type: str = Field(..., description="daily, weekly, monthly")


class ProviderPerformanceMetricsCreate(ProviderPerformanceMetricsBase):
    pass


class ProviderPerformanceMetrics(ProviderPerformanceMetricsBase):
    id: str
    user_id: int
    total_analyses: int = 0
    successful_analyses: int = 0
    failed_analyses: int = 0
    avg_processing_time: float | None = None
    total_cost: float | None = None
    avg_cost_per_analysis: float | None = None
    avg_response_length: float | None = None
    avg_token_usage: float | None = None
    user_satisfaction_score: float | None = None
    analysis_type_breakdown: dict[str, int] | None = None
    success_rate: float | None = None
    efficiency_score: float | None = None
    reliability_score: float | None = None
    calculated_at: datetime

    class Config:
        from_attributes = True


# Analysis Trend Schemas
class AnalysisTrendBase(BaseModel):
    trend_type: str = Field(..., description="health_improvement, recommendation_effectiveness, pattern_detection")
    analysis_type: str | None = Field(None, description="Filter by analysis type")
    metric_focus: str | None = Field(None, description="blood_pressure, weight, etc.")
    period_start: datetime
    period_end: datetime


class AnalysisTrendCreate(AnalysisTrendBase):
    pass


class AnalysisTrend(AnalysisTrendBase):
    id: str
    user_id: int
    data_points_count: int
    trend_direction: str | None = None  # improving, declining, stable, cyclical
    trend_strength: float | None = None  # Statistical significance (0-1)
    confidence_level: float | None = None  # Statistical confidence (0-1)
    correlation_coefficient: float | None = None
    regression_data: dict[str, Any] | None = None
    seasonal_patterns: dict[str, Any] | None = None
    outliers: list[dict[str, Any]] | None = None
    key_insights: list[str] | None = None
    improvement_suggestions: list[str] | None = None
    next_analysis_suggestions: list[str] | None = None
    improvement_percentage: float | None = None
    rate_of_change: float | None = None
    calculated_at: datetime
    last_updated: datetime
    is_significant: bool = False

    class Config:
        from_attributes = True


# Analysis Improvement Suggestion Schemas
class AnalysisImprovementSuggestionBase(BaseModel):
    suggestion_type: str = Field(..., description="data_collection, analysis_frequency, provider_selection, context_improvement")
    priority_level: str = Field(..., description="high, medium, low")
    category: str = Field(..., description="accuracy, completeness, efficiency, health_outcomes")
    title: str
    description: str
    detailed_explanation: str | None = None
    action_steps: list[str] | None = None


class AnalysisImprovementSuggestionCreate(AnalysisImprovementSuggestionBase):
    analysis_id: int | None = None


class AnalysisImprovementSuggestionUpdate(BaseModel):
    status: str | None = Field(None, description="pending, acknowledged, implemented, dismissed")
    user_feedback: str | None = None
    implementation_notes: str | None = None
    effectiveness_score: float | None = None


class AnalysisImprovementSuggestion(AnalysisImprovementSuggestionBase):
    id: str
    user_id: int
    analysis_id: int | None = None
    expected_improvement: str | None = None
    effort_level: str | None = None  # low, medium, high
    implementation_time: str | None = None
    supporting_data: dict[str, Any] | None = None
    related_analyses: list[int] | None = None
    success_examples: list[dict[str, Any]] | None = None
    status: str = "pending"
    user_feedback: str | None = None
    implementation_notes: str | None = None
    implemented_at: datetime | None = None
    effectiveness_score: float | None = None
    follow_up_analysis_ids: list[int] | None = None
    created_at: datetime
    updated_at: datetime
    expires_at: datetime | None = None

    class Config:
        from_attributes = True


# Comparison-specific response schemas
class ComparisonResult(BaseModel):
    """Result of comparing two or more analyses"""
    comparison_id: str
    analysis_ids: list[int]
    comparison_type: str
    summary: dict[str, Any]
    differences: list[dict[str, Any]]
    similarities: list[dict[str, Any]]
    statistical_analysis: dict[str, Any] | None = None
    recommendations: list[str]
    confidence_score: float | None = None


class TrendAnalysisResult(BaseModel):
    """Result of trending analysis over time"""
    trend_id: str
    analysis_ids: list[int]
    time_period: dict[str, datetime]
    trend_direction: str
    trend_strength: float
    key_insights: list[str]
    statistical_data: dict[str, Any]
    projections: dict[str, Any] | None = None
    improvement_metrics: dict[str, Any] | None = None


class ProviderComparisonResult(BaseModel):
    """Result comparing AI provider performance"""
    providers: list[dict[str, Any]]
    comparison_period: dict[str, datetime]
    metrics: dict[str, dict[str, float]]  # provider_id -> metric_name -> value
    rankings: dict[str, list[str]]  # metric_name -> ordered list of provider_ids
    recommendations: dict[str, str]  # recommendation_type -> provider_id
    cost_analysis: dict[str, Any]
    efficiency_analysis: dict[str, Any]


# Request schemas for comparison operations
class ComparisonRequest(BaseModel):
    analysis_ids: list[int] = Field(..., min_items=2, description="At least 2 analyses to compare")
    comparison_type: str = Field(..., description="side_by_side, temporal_trend, provider_performance")
    include_statistical_analysis: bool = Field(True, description="Include statistical analysis")
    focus_areas: list[str] | None = Field(None, description="Specific areas to focus comparison on")
    save_comparison: bool = Field(False, description="Save this comparison for future reference")
    comparison_name: str | None = Field(None, description="Name for saved comparison")


class TrendAnalysisRequest(BaseModel):
    analysis_type: str | None = Field(None, description="Filter by specific analysis type")
    metric_focus: str | None = Field(None, description="Focus on specific health metric")
    period_start: datetime
    period_end: datetime
    min_data_points: int = Field(3, description="Minimum number of analyses required")
    include_projections: bool = Field(True, description="Include future projections")
    statistical_tests: list[str] = Field(["correlation", "regression"], description="Statistical tests to perform")


class ProviderPerformanceRequest(BaseModel):
    provider_ids: list[str] | None = Field(None, description="Specific providers to analyze, or all if None")
    period_start: datetime
    period_end: datetime
    metrics: list[str] = Field(["success_rate", "avg_processing_time", "cost_efficiency"], description="Metrics to include")
    include_recommendations: bool = Field(True, description="Include provider recommendations")
    analysis_type_filter: str | None = Field(None, description="Filter by analysis type")

File: frontend/src/services/analysisComparison.ts
import api from './api';

// Comparison Types
export interface ComparisonRequest {
  analysis_ids: number[];
  comparison_type: 'side_by_side' | 'temporal_trend' | 'provider_performance';
  include_statistical_analysis?: boolean;
  focus_areas?: string[];
  save_comparison?: boolean;
  comparison_name?: string;
}

export interface ComparisonResult {
  comparison_id: string;
  analysis_ids: number[];
  comparison_type: string;
  summary: Record<string, any>;
  differences: ComparisonDifference[];
  similarities: ComparisonSimilarity[];
  statistical_analysis?: Record<string, any>;
  recommendations: string[];
  confidence_score?: number;
}

export interface ComparisonDifference {
  type: string;
  description: string;
  severity: 'low' | 'medium' | 'high';
}

export interface ComparisonSimilarity {
  type: string;
  description: string;
  strength: 'low' | 'medium' | 'high';
}

// Trend Analysis Types
export interface TrendAnalysisRequest {
  analysis_type?: string;
  metric_focus?: string;
  period_start: string;
  period_end: string;
  min_data_points?: number;
  include_projections?: boolean;
  statistical_tests?: string[];
}

export interface TrendAnalysisResult {
  trend_id: string;
  analysis_ids: number[];
  time_period: { start: string; end: string };
  trend_direction: string;
  trend_strength: number;
  key_insights: string[];
  statistical_data: Record<string, any>;
  projections?: Record<string, any>;
  improvement_metrics?: Record<string, any>;
}

// Provider Performance Types
export interface ProviderPerformanceRequest {
  provider_ids?: string[];
  period_start: string;
  period_end: string;
  metrics?: string[];
  include_recommendations?: boolean;
  analysis_type_filter?: string;
}

export interface ProviderComparisonResult {
  providers: ProviderInfo[];
  comparison_period: { start: string; end: string };
  metrics: Record<string, Record<string, number>>;
  rankings: Record<string, string[]>;
  recommendations: Record<string, string>;
  cost_analysis: Record<string, any>;
  efficiency_analysis: Record<string, any>;
}

export interface ProviderInfo {
  name: string;
  analysis_count: number;
}

// Saved Comparison Types
export interface AnalysisComparison {
  id: string;
  user_id: number;
  name: string;
  description?: string;
  analysis_ids: number[];
  comparison_type: string;
  comparison_criteria?: Record<string, any>;
  comparison_results?: Record<string, any>;
  key_differences?: Record<string, any>;
  statistical_insights?: Record<string, any>;
  trend_analysis?: Record<string, any>;
  created_at: string;
  updated_at: string;
  is_shared: boolean;
}

export interface AnalysisComparisonCreate {
  name: string;
  description?: string;
  analysis_ids: number[];
  comparison_type: string;
  comparison_criteria?: Record<string, any>;
}

export interface AnalysisComparisonUpdate {
  name?: string;
  description?: string;
  comparison_criteria?: Record<string, any>;
  is_shared?: boolean;
}

// Provider Performance Metrics Types
export interface ProviderPerformanceMetrics {
  id: string;
  provider_id: string;
  user_id: number;
  period_start: string;
  period_end: string;
  period_type: string;
  total_analyses: number;
  successful_analyses: number;
  failed_analyses: number;
  avg_processing_time?: number;
  total_cost?: number;
  avg_cost_per_analysis?: number;
  avg_response_length?: number;
  avg_token_usage?: number;
  user_satisfaction_score?: number;
  analysis_type_breakdown?: Record<string, number>;
  success_rate?: number;
  efficiency_score?: number;
  reliability_score?: number;
  calculated_at: string;
}

// Analysis Trend Types
export interface AnalysisTrend {
  id: string;
  user_id: number;
  trend_type: string;
  analysis_type?: string;
  metric_focus?: string;
  period_start: string;
  period_end: string;
  data_points_count: number;
  trend_direction?: string;
  trend_strength?: number;
  confidence_level?: number;
  correlation_coefficient?: number;
  regression_data?: Record<string, any>;
  seasonal_patterns?: Record<string, any>;
  outliers?: Record<string, any>[];
  key_insights?: string[];
  improvement_suggestions?: string[];
  next_analysis_suggestions?: string[];
  improvement_percentage?: number;
  rate_of_change?: number;
  calculated_at: string;
  last_updated: string;
  is_significant: boolean;
}

// Improvement Suggestion Types
export interface AnalysisImprovementSuggestion {
  id: string;
  user_id: number;
  analysis_id?: number;
  suggestion_type: string;
  priority_level: 'high' | 'medium' | 'low';
  category: string;
  title: string;
  description: string;
  detailed_explanation?: string;
  action_steps?: string[];
  expected_improvement?: string;
  effort_level?: 'low' | 'medium' | 'high';
  implementation_time?: string;
  supporting_data?: Record<string, any>;
  related_analyses?: number[];
  success_examples?: Record<string, any>[];
  status: 'pending' | 'acknowledged' | 'implemented' | 'dismissed';
  user_feedback?: string;
  implementation_notes?: string;
  implemented_at?: string;
  effectiveness_score?: number;
  follow_up_analysis_ids?: number[];
  created_at: string;
  updated_at: string;
  expires_at?: string;
}

export interface AnalysisImprovementSuggestionCreate {
  analysis_id?: number;
  suggestion_type: string;
  priority_level: 'high' | 'medium' | 'low';
  category: string;
  title: string;
  description: string;
  detailed_explanation?: string;
  action_steps?: string[];
}

export interface AnalysisImprovementSuggestionUpdate {
  status?: 'pending' | 'acknowledged' | 'implemented' | 'dismissed';
  user_feedback?: string;
  implementation_notes?: string;
  effectiveness_score?: number;
}

class AnalysisComparisonService {
  // Real-time Comparison Methods
  async compareAnalyses(request: ComparisonRequest): Promise<ComparisonResult> {
    const response = await api.post('/analysis-comparison/compare', request);
    return response.data;
  }

  async analyzeTrends(request: TrendAnalysisRequest): Promise<TrendAnalysisResult> {
    const response = await api.post('/analysis-comparison/trends/analyze', request);
    return response.data;
  }

  async analyzeProviderPerformance(request: ProviderPerformanceRequest): Promise<ProviderComparisonResult> {
    const response = await api.post('/analysis-comparison/providers/performance', request);
    return response.data;
  }

  // Saved Comparisons Management
  async createComparison(comparison: AnalysisComparisonCreate): Promise<AnalysisComparison> {
    const response = await api.post('/analysis-comparison/', comparison);
    return response.data;
  }

  async getComparisons(limit: number = 50): Promise<AnalysisComparison[]> {
    const response = await api.get('/analysis-comparison/', { params: { limit } });
    return response.data;
  }

  async getComparison(comparisonId: string): Promise<AnalysisComparison> {
    const response = await api.get(`/analysis-comparison/${comparisonId}`);
    return response.data;
  }

  async updateComparison(comparisonId: string, update: AnalysisComparisonUpdate): Promise<AnalysisComparison> {
    const response = await api.put(`/analysis-comparison/${comparisonId}`, update);
    return response.data;
  }

  async deleteComparison(comparisonId: string): Promise<void> {
    await api.delete(`/analysis-comparison/${comparisonId}`);
  }

  // Provider Performance Metrics
  async getProviderMetrics(
    providerId?: string,
    periodType: string = 'monthly',
    limit: number = 12
  ): Promise<ProviderPerformanceMetrics[]> {
    const params: any = { period_type: periodType, limit };
    if (providerId) params.provider_id = providerId;
    
    const response = await api.get('/analysis-comparison/providers/metrics', { params });
    return response.data;
  }

  // Analysis Trends
  async getAnalysisTrends(
    trendType?: string,
    analysisType?: string,
    significantOnly: boolean = false,
    limit: number = 20
  ): Promise<AnalysisTrend[]> {
    const params: any = { significant_only: significantOnly, limit };
    if (trendType) params.trend_type = trendType;
    if (analysisType) params.analysis_type = analysisType;
    
    const response = await api.get('/analysis-comparison/trends/', { params });
    return response.data;
  }

  // Improvement Suggestions
  async createImprovementSuggestion(suggestion: AnalysisImprovementSuggestionCreate): Promise<AnalysisImprovementSuggestion> {
    const response = await api.post('/analysis-comparison/suggestions/', suggestion);
    return response.data;
  }

  async getImprovementSuggestions(
    statusFilter?: string,
    priorityFilter?: string,
    categoryFilter?: string,
    limit: number = 50
  ): Promise<AnalysisImprovementSuggestion[]> {
    const params: any = { limit };
    if (statusFilter) params.status_filter = statusFilter;
    if (priorityFilter) params.priority_filter = priorityFilter;
    if (categoryFilter) params.category_filter = categoryFilter;
    
    const response = await api.get('/analysis-comparison/suggestions/', { params });
    return response.data;
  }

  async updateImprovementSuggestion(
    suggestionId: string, 
    update: AnalysisImprovementSuggestionUpdate
  ): Promise<AnalysisImprovementSuggestion> {
    const response = await api.put(`/analysis-comparison/suggestions/${suggestionId}`, update);
    return response.data;
  }

  // Analytics Summary
  async getAnalyticsSummary(): Promise<{
    total_comparisons: number;
    significant_trends: number;
    pending_suggestions: number;
    recent_provider_metrics: number;
    has_comparison_data: boolean;
  }> {
    const response = await api.get('/analysis-comparison/analytics/summary');
    return response.data;
  }
}

export const analysisComparisonService = new AnalysisComparisonService();